src/libos/src/time/timer_file.rs

```diff
@@ -0,0 +1,387 @@
+                                // Error or Stopped by tirq
+                                break;
+                            } else if one_shot == true {
+                                // One-shot timer, no need loop
+                                trace!("timerfd one-shot triggerred");
+                                expired_closure(&inner_c, true);
+                                break;
+                            } else {
+                                trace!("timerfd timer expired");
+                                // If it is not one-shot, the following timeout changes to interval.
+                                timeout = interval;
+                                expired_closure(&inner_c, false);
+                            }
+                        }
+                        _ => {
+                            panic!("impossible as there is no waker or timeout");
+                        }
+                    }
+                }
+
+                trace!("Timerfd poll task end");
+            }
+        })
+        .priority(async_rt::sched::SchedPriority::Low)
+        .spawn();
+
+        inner.status = TimerFdStatus::RUN;
+        inner.task_handle = Some(join_handle);
+
+        Ok(())
+    }
+
+    async fn wait_for_expire(&self) -> Result<usize> {
+        let flags = self.flags.load(Ordering::Relaxed);
+        let mut count: usize = 0;
+
+        let inner_c = self.inner.clone();
+        let mut inner = inner_c.lock().unwrap();
+        if inner.status == TimerFdStatus::STOP {
+            return_errno!(EAGAIN, "try again");
+        } else if inner.status == TimerFdStatus::EXPIRE {
+            count = inner.exp_cnt;
+            return Ok(count);
+        }
+
+        // Do not hold the Mutex for below whole loop
+        drop(inner);
+
+        let exp_waiters = &self.exp_waiters.clone();
+        waiter_loop!(exp_waiters, {
+            let mut inner = inner_c.lock().unwrap();
+            count = inner.exp_cnt;
+            drop(inner);
+
+            if count > 0 {
+                return Ok(count);
+            } else if flags.contains(TimerCreationFlags::TFD_NONBLOCK) {
+                return_errno!(EAGAIN, "try again");
+            }
+        })
+        .map_err(|_| errno!(EAGAIN, "try again"))
+    }
+}
+
+bitflags! {
+    pub struct TimerCreationFlags: i32 {
+        /// Provides semaphore-like semantics for reads from the new file descriptor
+        /// Non-blocking
+        const TFD_NONBLOCK  = 1 << 11;
+        /// Close on exec
+        const TFD_CLOEXEC   = 1 << 19;
+    }
+}
+
+bitflags! {
+    pub struct TimerSetFlags: i32 {
+        const TFD_TIMER_ABSTIME = 1 << 0;
+        const TFD_TIMER_CANCEL_ON_SET = 1 << 1;
+    }
+}
+
+impl TimerFile {
+    pub async fn read(&self, buf: &mut [u8]) -> Result<usize> {
+        if buf.len() < 8 {
+            return_errno!(EINVAL, "buffer is too small");
+        }
+
+        let count = self.wait_for_expire().await?;
```

Comment:
The inner.exp_cnt may be increased after we get the count by let count = self.wait_for_expire().await, the line "inner.exp_cnt = 0" reset the count may miss some counts by mistake.

Reply:
You are right. Do not reset it to 0 but reduce the expired count alread read could avoid the race.
```
        let mut inner = inner_c.lock().unwrap();
        // Reduce expired count alread read
        inner.exp_cnt = inner.exp_cnt - count;
```